src/components/vintage/vintage.cairo

```diff
@@ -0,0 +1,233 @@
+#[starknet::component]
+mod VintageComponent {
+    // Starknet imports
+    use starknet::{get_block_timestamp, get_caller_address};
+
+    // Internal imports
+    use carbon_v3::components::vintage::interface::IVintage;
+    use carbon_v3::models::carbon_vintage::{CarbonVintage, CarbonVintageType};
+
+    // Constants
+    use carbon_v3::models::constants::{CC_DECIMALS, CC_DECIMALS_MULTIPLIER};
+    use carbon_v3::contracts::project::Project::OWNER_ROLE;
+
+    // Roles
+    use openzeppelin::access::accesscontrol::interface::IAccessControl;
+
+    #[storage]
+    struct Storage {
+        Vintage_vintages: LegacyMap<u256, CarbonVintage>,
+        Vintage_vintages_len: usize,
+        Vintage_project_carbon: u128,
+    }
+
+    #[event]
+    #[derive(Drop, PartialEq, starknet::Event)]
+    enum Event {
+        ProjectCarbonUpdate: ProjectCarbonUpdate,
+        VintageUpdate: VintageUpdate,
+    }
+
+    #[derive(Drop, PartialEq, starknet::Event)]
+    struct ProjectCarbonUpdate {
+        old_carbon: u128,
+        new_carbon: u128,
+    }
+
+
+    #[derive(Drop, PartialEq, starknet::Event)]
+    struct VintageUpdate {
+        #[key]
+        token_id: u256,
+        vintage: CarbonVintage,
+    }
+
+    mod Errors {
+        const INVALID_ARRAY_LENGTH: felt252 = 'Absorber: invalid array length';
+        const INVALID_STARTING_YEAR: felt252 = 'Absorber: invalid starting year';
+    }
+
+    #[embeddable_as(VintageImpl)]
+    impl Vintage<
+        TContractState,
+        +HasComponent<TContractState>,
+        +Drop<TContractState>,
+        +IAccessControl<TContractState>
+    > of IVintage<ComponentState<TContractState>> {
+        fn get_project_carbon(self: @ComponentState<TContractState>) -> u128 {
+            let mut project_carbon: u128 = 0;
+
+            // let mut index = self.Vintage_vintages_len.read();
+            // while index > 0 {
+            //     index -= 1;
+            //     let vintage = self.Vintage_vintages.read(index.into());
+            //     if vintage.status != CarbonVintageType::Unset {
+            //         project_carbon += vintage.supply.into();
+            //     }
+            // };
```

Comment:
Why not calculating `Vintage_project_carbon` as you did here, and remove the storage var? However we should change that storage value when we update a Vintage with `rebase_vintage` or `rebase_vintage`

Reply:
indeed, we still need to figure out if we need this storage var.
The issue is to make it work with the whole process: minting, updating the project carbon before minting and rebasing.
This implementation might not be the final one.